### PR TITLE
[WorkGraphs] Proposal to Add MaxRecordsPerNode attribute to NodeOutputArrays.

### PR DESCRIPTION
**Pull request by anupamachandra**
WRT Workgraphs Feature:
This is the HLSL cpunterpart of the proposal to new node output attribute called `MaxRecordsPerNode`. This parameter is only required for output node arrays.  This attribute specifies the maximum number of records that can be written to any single output node within a node array.
When determining backing store memory requirements, an implementation must assume the worst-case of `MaxRecords` written
to any single node in the output array.  However, a common use-case is for a small number records to be written to
select nodes in a very large array of nodes.  Some implementations can take advantage of this knowledge to significantly
reduce the backing store memory requirements while maintaining peak performance.

### CODE REVIEW

**Comment by anupamachandra** [comment]
@microsoft-github-policy-service agree company="Nvidia"

File: proposals/NNNN-max-records-per-node.md
Comment (tex3d) [inline proposals/NNNN-max-records-per-node.md:47]: The way to do this is to use a warning that defaults to error (a `DefaultError` warning) with a warning group that can be addressed at the command line to override this default (change to warning or silence).

```suggestion
Since this may cause compilation failures with existing Work Graphs, this will
be a `DefaultError` warning assigned to a warning group named
`hlsl-require-max-records-per-node` to allow a command-line override.
The value of `MaxRecordsPerNode` will be set equal to `MaxRecords`.
```

File: proposals/NNNN-max-records-per-node.md
Comment (tex3d) [inline proposals/NNNN-max-records-per-node.md:100]: `MaxRecordsPerNode` is defined as a separate attribute earlier on, instead of an extra parameter to the existing `MaxRecords` attribute.  Which is it?

File: proposals/NNNN-max-records-per-node.md
Comment (tex3d) [inline proposals/NNNN-max-records-per-node.md:112]: This case is already described above to have a different behavior.

File: proposals/NNNN-max-records-per-node.md
Comment (mapodaca-nv) [reply proposals/NNNN-max-records-per-node.md:100]: The proposal is a separate attribute.  This section shows an "alternative" method as an extra parameter, which was earlier discussed in other forums.

Comment (anupamachandra) [reply proposals/NNNN-max-records-per-node.md]: @tex3d Better?

Comment (tex3d) [reply proposals/NNNN-max-records-per-node.md]: Apologies for the confusion from before.  For some reason when I was reading these I wasn't realizing they were under an "alternative" section.

File: proposals/NNNN-max-records-per-node.md
Comment (tex3d) [inline proposals/NNNN-max-records-per-node.md:5]: Before we merge the PR, we need to find a `Sponsor`.  You'll want to update the number `NNNN` to the next number just before merging as well.

**Review by tex3d — APPROVED** [review]
I think this is fine to merge once we find a sponsor and update the document number `NNNN`.

File: proposals/NNNN-max-records-per-node.md
Comment (anupamachandra) [reply proposals/NNNN-max-records-per-node.md:5]: What's the process to get a Sponsor assigned?

Comment (damyanp) [reply proposals/NNNN-max-records-per-node.md:5]: Feel free to put me down as sponsor.

File: proposals/NNNN-max-records-per-node.md
Comment (damyanp) [inline proposals/NNNN-max-records-per-node.md:3]: You also need to rename the file (it is currently `proposals/NNN-max-records.per-node.md`) and make the link on this line contain the correct filename.

File: proposals/0025-max-records-per-node.md
Comment (damyanp) [inline proposals/0025-max-records-per-node.md:3]:
```suggestion
* Proposal: [0025](0025-max-records-per-node.md)
```

Comment (anupamachandra) [reply proposals/0025-max-records-per-node.md]: Done!

**Comment by anupamachandra** [comment]
@damyanp Is anything else required from my end?

**Review by damyanp — APPROVED** [review]

**Pull request merged**
